### ISSUE TITLE
NodeAnimation exporter

### ISSUE THREAD

**Issue by DarioDaF** [OGRECave/blender2ogre#58]
I started using Ogre and I need many simple animations, translation-rotation keyframes for objects.
In blender I used the standard animation keyframe giving names to animations as seen in the image:
![ExportNodeAnim](https://user-images.githubusercontent.com/10617344/100517070-7305b800-3188-11eb-96ed-4861df396035.png)
Unfortunately the exporter supports only skeletal animation, but I found that http://wiki.ogre3d.org/Easy+Ogre+Exporter for 3DS Max supports them https://github.com/OGRECave/EasyOgreExporter/blob/629da70741f9bcad18466868cb3631cffdefe7f4/source/ExScene.cpp#L104
I don't know much about the blender side of the plugin, but it seems 
https://github.com/OGRECave/ogre/blob/master/Tools/XMLConverter/docs/ogremeshxml.dtd
does not allow dumping ketframes directly under animation nodes so I would like to know if it simply is not developed yet or there are some Ogre version differences.

**Comment by paroj** [comment]
> Unfortunately the exporter supports only skeletal animation, but I found that http://wiki.ogre3d.org/Easy+Ogre+Exporter for 3DS Max supports them https://github.com/OGRECave/EasyOgreExporter/blob/629da70741f9bcad18466868cb3631cffdefe7f4/source/ExScene.cpp#L104

this is a non-standard node, that EOE is exporting there for .scene 

> I don't know much about the blender side of the plugin, but it seems
https://github.com/OGRECave/ogre/blob/master/Tools/XMLConverter/docs/ogremeshxml.dtd
does not allow dumping ketframes directly under animation nodes

you are looking at the wrong DTD there. The .mesh only stores geometry (and bone weights) and geometry based keyframes (aka Morph animation).

the correct one for node animations would be:
https://github.com/OGRECave/ogre/blob/master/PlugIns/DotScene/misc/dotscene.dtd

However, it does not specify node animations either and there is no code to load them in the plugin, so even with EOE that would need to be added.
Actually that would be quite straightforward - see e.g.
https://github.com/OGRECave/ogre/blob/c0dea00db6036e40f5eb1dad3714358f1d1762b1/Samples/Simple/include/CameraTrack.h#L43-L59

> I would like to know if it simply is not developed yet or there are some Ogre version differences.

Basically it simply is not developed yet - one could just add the EOE format to the .scene DTD and add the missing bits to blender2ogre and the plugin.

**Comment by sercero** [comment]
@DarioDaF is there a reason you can't use skeleton based animations?

**Comment by DarioDaF** [comment]
In practice I used another library instead of ogre, but the problem is both of clarity (skin association etc are trivial and cannot be easily checked in code) and also because skeletal animation in blender requires bones and mesh application, so it would have slowed a bit (a lot considering there are many small animations) the 3d modeller (unless creating anyway a custom importer to convert the plain animation to skeletal)

**Comment by sercero** [comment]
@paroj in the example provided here:
https://github.com/OGRECave/ogre/blob/c0dea00db6036e40f5eb1dad3714358f1d1762b1/Samples/Simple/include/CameraTrack.h#L43-L59

```
track->createNodeKeyFrame(0)->setTranslate(Vector3(200, 0, 0));
```

In the manual it says that `KeyFrame(0)` should already exist and you should `get()` it:
https://ogrecave.github.io/ogre/api/1.12/class_ogre_1_1_node_animation_track.html#a5018654d95b4bca786c90c073fba4898

Do you know if that code generates any trouble?

I'm asking because I am in the process of implementing this in `blender2ogre` and later in `DotScene`.

**Comment by paroj** [comment]
just stepped with a debugger into that. The manual is wrong. There is no KeyFrame(0)
